CRM/Price/BAO/LineItem.php

```diff
@@ -605,4 +604,558 @@ public static function getLineItemArray(&$params, $entityId = NULL, $entityTable
     }
   }
 
+  /**
+   * Function to update related contribution of a entity and
+   *  add/update/cancel financial records
+   *
+   * @param array $params
+   * @param int $entityID
+   * @param int $entity
+   * @param int $contributionId
+   * @param $feeBlock
+   * @param array $lineItems
+   * @param $paidAmount
+   *
+   */
+  public static function changeFeeSelections(
+    $params,
+    $entityID,
+    $entity,
+    $contributionId,
+    $feeBlock,
+    $lineItems,
+    $paidAmount
+  ) {
+    $entityTable = "civicrm_" . $entity;
+    CRM_Price_BAO_PriceSet::processAmount($feeBlock,
+      $params, $lineItems
+    );
+    // initialize empty Lineitem instance to call protected helper functions
+    $lineItemObj = new CRM_Price_BAO_LineItem();
+
+    // fetch submitted LineItems from input params and feeBlock information
+    $submittedLineItems = $lineItemObj->_getSubmittedLineItems($params, $feeBlock);
+
+    // retrieve the submitted price field value IDs from $submittedLineItems array keys
+    $submittedPriceFieldValueIDs = empty($submittedLineItems) ? array() : array_keys($submittedLineItems);
+
+    // get lineItems need to be updated and added to record changed fee
+    list($lineItemsToAdd, $lineItemsToUpdate) = $lineItemObj->_getLineItemsToAddAndUpdate($submittedLineItems, $entityID, $entity);
+
+    // cancel previous line item
+    $additionalWhereClause = empty($submittedPriceFieldValueIDs) ? NULL : sprintf("price_field_value_id NOT IN (%s)", implode(', ', $submittedPriceFieldValueIDs));
+    $lineItemObj->_cancelLineItems($entityID, $entityTable, $additionalWhereClause);
+
+    // get financial information that need to be recorded on basis on submitted price field value IDs
+    $financialItemsArray = $lineItemObj->_getFinancialItemsToRecord(
+      $entityID,
+      $entityTable,
+      $contributionId,
+      $submittedPriceFieldValueIDs
+    );
+
+    // update line item with changed line total and other information
+    $totalParticipant = $participantCount = 0;
+    $amountLevel = array();
+    if (!empty($lineItemsToUpdate)) {
+      foreach ($lineItemsToUpdate as $priceFieldValueID => $value) {
+        $taxAmount = "NULL";
+        if (isset($value['tax_amount'])) {
+          $taxAmount = $value['tax_amount'];
+        }
+        $amountLevel[] = $value['label'] . ' - ' . (float) $value['qty'];
+        if ($entity == 'participant' && isset($value['participant_count'])) {
+          $participantCount = $value['participant_count'];
+          $totalParticipant += $value['participant_count'];
+        }
+        $updateLineItemSQL = "
+        UPDATE civicrm_line_item li
+        SET li.qty = {$value['qty']},
```

Comment:
@seamuslee001 to ensure that I have put a ```!empty(...)``` check on above and once there is a ```$lineItemsToUpdate``` it will always contain necessary attributes. Also, this has been already tested earlier during ```participant-change-fee-selection```. See https://github.com/civicrm/civicrm-core/pull/10370/files#diff-d05dcb57339ac128d06193a57b9ca1bcL2024